extensions/ads-windows-support/src/telemetry.ts

```diff
@@ -0,0 +1,136 @@
+     */
+    public static initialize(): void {
+        if (typeof this.reporter === 'undefined') {
+            // Check if the user has opted out of telemetry
+            if (!vscode.workspace.getConfiguration('telemetry').get<boolean>('enableTelemetry', true)) {
+                this.disable();
+                return;
+            }
+
+            let packageInfo = Utils.getPackageInfo(packageJson);
+            this.reporter = new TelemetryReporter(packageInfo.name, packageInfo.version, packageInfo.aiKey);
+        }
+    }
+
+    /**
+     * Send a telemetry event for an exception
+     */
+    public static sendTelemetryEventForException(
+        err: any, methodName: string, extensionConfigName: string): void {
+        try {
+            let stackArray: string[];
+            let firstLine: string = '';
+            if (err !== undefined && err.stack !== undefined) {
+                stackArray = err.stack.split('\n');
+                if (stackArray !== undefined && stackArray.length >= 2) {
+                    firstLine = stackArray[1]; // The first line is the error message and we don't want to send that telemetry event
+                    firstLine = FilterErrorPath(firstLine);
+                }
+            }
+
+            // Only adding the method name and the fist line of the stack trace. We don't add the error message because it might have PII
+            this.sendTelemetryEvent('Exception', { methodName: methodName, errorLine: firstLine });
+            // Utils.logDebug('Unhandled Exception occurred. error: ' + err + ' method: ' + methodName, extensionConfigName);
+        } catch (telemetryErr) {
+            // If sending telemetry event fails ignore it so it won't break the extension
+            // Utils.logDebug('Failed to send telemetry event. error: ' + telemetryErr, extensionConfigName);
+        }
+    }
+
+    /**
+     * Send a telemetry event using application insights
+     */
+    public static sendTelemetryEvent(
+        eventName: string,
+        properties?: ITelemetryEventProperties,
+        measures?: ITelemetryEventMeasures): void {
+
+        if (typeof this.disabled === 'undefined') {
+            this.disabled = false;
+        }
+
+        if (this.disabled || typeof (this.reporter) === 'undefined') {
+            // Don't do anything if telemetry is disabled
+            return;
+        }
+
+        if (!properties || typeof properties === 'undefined') {
+            properties = {};
+        }
+
+        // Augment the properties structure with additional common properties before sending
+        Promise.all([this.getPlatformInformation()]).then(() => {
+            properties['distribution'] = (this.platformInformation && this.platformInformation.distribution) ?
+                `${this.platformInformation.distribution.name}, ${this.platformInformation.distribution.version}` : '';
+
+            this.reporter.sendTelemetryEvent(eventName, properties, measures);
```

Comment:
I don't see the extension name being added to the properties, if not, we should definitely add it.